src/dotnet-grpc/Commands/ListCommand.cs

```diff
@@ -0,0 +1,102 @@
+using System.Linq;
+using Grpc.Dotnet.Cli.Internal;
+using Grpc.Dotnet.Cli.Options;
+using Grpc.Dotnet.Cli.Properties;
+using Microsoft.Build.Evaluation;
+
+namespace Grpc.Dotnet.Cli.Commands
+{
+    internal class ListCommand : CommandBase
+    {
+        public ListCommand(IConsole console, FileInfo? projectPath)
+            : base(console, projectPath) { }
+
+        public static Command Create()
+        {
+            var command = new Command(
+                name: "list",
+                description: CoreStrings.ListCommandDescription);
+            command.AddOption(CommonOptions.ProjectOption());
+
+            command.Handler = CommandHandler.Create<IConsole, FileInfo>(
+                (console, project) =>
+                {
+                    try
+                    {
+                        var command = new ListCommand(console, project);
+                        command.List();
+
+                        return 0;
+                    }
+                    catch (CLIToolException e)
+                    {
+                        console.LogError(e);
+
+                        return -1;
+                    }
+                });
+
+            return command;
+        }
+
+        public  void List()
+        {
+            var consoleRenderer = new ConsoleRenderer(Console);
+            var protobufElements = Project.GetItems(ProtobufElement).ToList();
+            var table = new TableView<ProjectItem> { Items = protobufElements};
+
+            // Required columns (always displayed)
+            table.AddColumn(r => r.UnevaluatedInclude, "Protobuf Reference");
+            table.AddColumn(r =>
+            {
+                var serviceType = r.GetMetadataValue(GrpcServicesElement);
+                return string.IsNullOrEmpty(serviceType) ? "Both" : serviceType;
+            }, "Service Type");
+
+            // Optional columns (only displayed if an element is not default)
+            if (protobufElements.Any(r => !string.IsNullOrEmpty(r.GetMetadataValue(SourceUrlElement))))
+            {
+                table.AddColumn(r => r.GetMetadataValue(SourceUrlElement), "Source URL");
+            }
+
+            // The default value is Public set by Grpc.Tools so skip this column if everything is default
+            if (protobufElements.Any(r => !string.Equals(r.GetMetadataValue(AccessElement), Access.Public.ToString(), StringComparison.OrdinalIgnoreCase)))
+            {
+                table.AddColumn(r => r.GetMetadataValue(AccessElement), "Access");
+            }
+
+            if (protobufElements.Any(r => !string.IsNullOrEmpty(r.GetMetadataValue(AdditionalImportDirsElement))))
+            {
+                table.AddColumn(r => r.GetMetadataValue(AdditionalImportDirsElement), "Additional Imports");
+            }
+
+            var screen = new ScreenView(consoleRenderer, Console) { Child = table };
+            screen.Render(new Region(0, 0, System.Console.WindowWidth, System.Console.WindowWidth));
```

Comment:
Fancy. Have you tested out what happens when file names/URLs are very long? Does it still look ok?

Reply:
Not great. This is mostly dependent on the TableView API which allows size to content (default), fixed, and star options. The other option is to use star and specify the weight (ratio based, e.g. 1:0.5:1) but the size is no longer contextual to the content and will cut off the content if it's too long:
```Protobuf Reference                                                                         Service Type                                                                               Source URL                                                                                 Access
..\Proto\aggregate.proto                                                                   Server                                                                                     Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed gravida odio nisi, sed laoreetInternal
..\Proto\certify.proto                                                                     Server                                                                                                                                                                                Public
..\Proto\count.proto                                                                       Both                                                                                                                                                                                  Public
..\Proto\greet.proto                                                                       Both                                                                                                                                                                                  Public
..\Proto\mail.proto                                                                        Server                                                                                                                                                                                Public
..\Proto\ticket.proto                                                                      Server                                                                                                                                                                                Public
```

Alternative is to calculate the column ourselves  and use ellipses when content is too long (which I didn't want to do). Could also consider asking for new features to make this better but I'm not sure about the turnaround time.